Faster identification of market indeces via pandas.Series.groupby

mcaceresb opened the pull request:
With a large number of markets, a large number of observations, or both, identifying the market indices in pyblp/economies/economy.py lines 76-80 becomes prohibitively time-consuming. We found that using pandas.Series.groupby got rid of the bottleneck.

PS: I wonder if maybe you didn't want pandas as a dependency? I figured that it might be OK to add it since all the docs assume the user will have pandas. The speedup seems to warrant it.

chrisconlon commented:
Jeff and I have discussed this before -- I'll leave the decision to him. We haven't done much in terms of low-level optimization yet. One I've used before that worked in pure Numpy was http://esantorella.com/2016/06/16/groupby/

mcaceresb commented:
Interesting. I don't think their benchmarks hold up, though. While pandas _is_ slower for user functions, as they say, it's actually much faster in the groupby statement. So you can get similar speeds for user-defined functions by simply calling the grouped indices directly. In their example you can do:

```python
grouped = df.groupby('first category')
result = np.array([np.mean(y[idx]) for idx in grouped.indices.values()])
```

and get a similar speed. However, their class is still slow for large N.

jeffgortmaker commented:
I haven't added pandas as a dependency exactly because (as you guessed) I can't bring myself to add a big dependency for a single function. Honestly this is sort of a philosophical reason on my part.

There are other places in the code where fast grouped operations are even more important (nesting structure, fixed effect absorption, etc.). Instead of adding pandas as a dependency, I've implemented grouped operations with a simple [Groups](https://github.com/jeffgortmaker/pyblp/blob/7f09893f4a4d5975efc061f918efd262e180a260/pyblp/utilities/basics.py#L320) class that looks similar to what Chris posted. I think its performance is similar to pandas, and doesn't have any overhead.

I didn't think to similarly speed up those two lines that are causing you trouble, but it shouldn't be difficult to speed them (and probably some similar places in the code) up with something that looks similar to `Groups.__init__`. If you want to update the PR to do something just with numpy, I'd really appreciate it! Otherwise I can fix this in the next couple of days. I think for now I'd rather have 50 lines of code instead of a pandas dependency unless pandas is doing some magic that I can't replicate with numpy.

mcaceresb commented:
I think the issue with argsort is that it will perform poorly with very large N. But it might be preferable than adding such a large dependency. Something like this I think:

```
def _indices(ids):
    sort_idx = ids.argsort(kind='mergesort')
    uniq_ids, sort_pos = np.unique(ids[sort_idx], return_index=True)
    return {k: v for k, v in zip(uniq_ids, np.split(sort_idx, sort_pos)[1:])}
```

maybe if a radix/counting sort is available that will match pandas' speed for grouping; the IDs are always integers right?

jeffgortmaker commented:
The IDs can be any object so that debugging is easier and printed output is more readable. I should probably require them to be just strings at some point for speed.

That looks great. Here's a modified version without the extra sort in `np.unique`:

    def indices(ids):
        flat = ids.flatten()
        sort_indices = flat.argsort(kind='mergesort')
        sorted_ids = flat[sort_indices]
        changes = np.ones(flat.shape, np.bool)
        changes[1:] = sorted_ids[1:] != sorted_ids[:-1]
        reduce_indices = np.nonzero(changes)[0]
        return dict(zip(sorted_ids[reduce_indices], np.split(sort_indices, reduce_indices)[1:]))

At least on strings, it seems to be pretty good compared to pandas, especially when markets are already sorted, which is typical for BLP datasets. Sorting a few million strings takes under 10 seconds, which isn't much compared to how long estimation will take.

If you want to take a shot at adding this, I'd put it in `utilities.basics` and replace those two lines along with [this one](https://github.com/jeffgortmaker/pyblp/blob/7f09893f4a4d5975efc061f918efd262e180a260/pyblp/construction.py#L357).

mcaceresb commented:
I had kind of lazily assumed np.unique would be more efficient if the input array was sorted.

Anyway, I removed the pandas dependency and modified the code so that it uses this function instead. As you point out, it will only be a few seconds slower in any reasonable use case so it shouldn't be a bottleneck anymore, and it doesn't add an unnecessary dependency. It seems like a reasonable solution.

jeffgortmaker commented on pyblp/economies/economy.py:
```diff
@@ -73,11 +73,8 @@ def __init__(
-            self._product_market_indices[t] = np.flatnonzero(self.products.market_ids == t)
-            self._agent_market_indices[t] = np.flatnonzero(self.agents.market_ids == t)
+        self._product_market_indices: Dict[Hashable, Array] = get_indices(self.products.market_ids)
+        self._agent_market_indices: Dict[Hashable, Array] = get_indices(self.agents.market_ids)
```
You can remove the type annotations here. They're now inferred by the return type of `get_indices`.

jeffgortmaker commented on pyblp/utilities/basics.py:
```diff
@@ -295,6 +295,16 @@ def format_table(
     return "\n".join(lines)
 
 
+def get_indices(ids) -> Dict[Hashable, Array]:
```
Can you add the type `Array` for `ids`. And also a docstring?

jeffgortmaker commented:
Great thanks! Added a couple comments.
